### ISSUE TITLE
Node status is now: CIDRNotAvailable

### ISSUE THREAD

**Issue by chuegel** [kubernetes/kubernetes#85866]
**What happened**:
We deployed a 5 node (3 master / 2 worker) cluster with rke:
```
curl https://docs.projectcalico.org/v3.10/getting-started/kubernetes/installation/hosted/kubernetes-datastore/calico-networking/1.7/calico.yaml |\
     sed 's,192.168.0.0/16,10.2.225.0/24,g' > calico_local.yaml

cat - > rke_cluster.yaml <<EOF
nodes:
  - address: 192.168.100.17
    user: chuegel
    ssh_key_path: ~/.ssh/id_rsa
    hostname_override: nas-10-control-03
    role: [controlplane,etcd]
  - address: 192.168.100.13
    user: chuegel
    ssh_key_path: ~/.ssh/id_rsa
    hostname_override: nas-10-control-01
    role: [controlplane,etcd]
  - address: 192.168.100.14
    user: chuegel
    ssh_key_path: ~/.ssh/id_rsa
    role: [controlplane,etcd]
    hostname_override: nas-10-control-02
  - address: 192.168.100.15
    user: chuegel
    ssh_key_path: ~/.ssh/id_rsa
    role: [worker]
    hostname_override: nas-10-worker-01
  - address: 192.168.100.16
    hostname_override: nas-10-worker-02
    user: chuegel
    ssh_key_path: ~/.ssh/id_rsa
    role: [worker]

services:
  etcd:
    snapshot: true
    creation: 6h
    retention: 24h
  kube-api:
    service_cluster_ip_range: 10.2.228.0/22
  kube-controller:
    service_cluster_ip_range: 10.2.228.0/22
    cluster_cidr: 10.2.225.0/24
  kubelet:
    cluster_domain: lab.local
    cluster_dns_server: 10.2.228.66

kubernetes_version: v1.16.3-rancher1-1

addon_job_timeout: 30

network:
    plugin: none

addons_include:
    - calico_local.yaml
EOF
```
I noticed this event repeating constantly:
```
7s          Normal    CIDRNotAvailable          node/nas-10-worker-02    Node nas-10-worker-02 status is now: CIDRNotAvailable
1s          Normal    CIDRNotAvailable          node/nas-10-control-01   Node nas-10-control-01 status is now: CIDRNotAvailable
7s          Normal   CIDRNotAvailable        node/nas-10-worker-01    Node nas-10-worker-01 status is now: CIDRNotAvailable
```
`kubectl cluster-info dump | grep -m 1 cluster-cidr`
is empty.

However, I can schedule pods on the nodes and get the right IP addresses and the cluster looks healthy:
```
calicoctl get wep --all-namespaces
NAMESPACE        WORKLOAD                                   NODE                NETWORKS          INTERFACE
cattle-system    cattle-cluster-agent-5549dddf78-7gbnx      nas-10-control-02   10.2.137.2/32     cali755fe14f303
cattle-system    rancher-c88c6458c-8xwht                    nas-10-worker-02    10.2.137.140/32   cali8139b4c3591
cattle-system    rancher-c88c6458c-m99kg                    nas-10-worker-01    10.2.137.130/32   caliba81a84bec4
cattle-system    rancher-c88c6458c-mccsj                    nas-10-worker-02    10.2.137.148/32   cali4deb49e4cb4
cert-manager     cert-manager-7b8cb8df9-t8ptz               nas-10-worker-01    10.2.136.29/32    cali1a1843fcb08
cert-manager     cert-manager-cainjector-6d86bc5b5c-wnlm9   nas-10-worker-02    10.2.136.31/32    cali0c118fbbcac
cert-manager     cert-manager-webhook-f67cf55c6-tnvw5       nas-10-worker-02    10.2.136.30/32    cali80c967334f5
idm-test         postgres-684ff44955-x8rr8                  nas-10-worker-02    10.2.137.146/32   cali78a01809603
ingress-nginx    default-http-backend-67cf578fc4-2mnlq      nas-10-worker-01    10.2.137.129/32   caliade5ead08be
kube-system      calico-kube-controllers-6b64bcd855-87zqc   nas-10-worker-01    10.2.137.135/32   calie4b42d898b8
kube-system      coredns-5c59fd465f-mrfmf                   nas-10-worker-02    10.2.137.144/32   calia99956f9d51
kube-system      coredns-5c59fd465f-p87x4                   nas-10-worker-01    10.2.137.138/32   calie7b6fed8cfb
kube-system      coredns-autoscaler-d765c8497-fjfs4         nas-10-worker-02    10.2.137.142/32   cali2ac676c4090
kube-system      metrics-server-64f6dffb84-7ctcn            nas-10-worker-01    10.2.137.145/32   cali9c88118fdab
kube-system      tiller-deploy-969865475-xclrb              nas-10-worker-02    10.2.137.143/32   cali749dbebbcfb
metallb-system   controller-65895b47d4-gmswr                nas-10-worker-02    10.2.137.147/32   cali9461a98b19f
```

**What you expected to happen**:
event not to occur
**How to reproduce it (as minimally and precisely as possible)**:
see above
**Anything else we need to know?**:
```
calicoctl version
Client Version:    v3.10.1
Git commit:        4aaff8e9
Cluster Version:   v3.10.1
Cluster Type:      k8s,bgp,kdd
```
```
calicoctl get ippool
NAME                  CIDR              SELECTOR
cert-manager          10.2.136.16/28    all()
default-ipv4-ippool   10.2.137.0/24     all()
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:23:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:13:49Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: VMs
- OS (e.g: `cat /etc/os-release`):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`):
`Linux nas-10-control01 3.10.0-1062.4.3.el7.x86_64 #1 SMP Wed Nov 13 23:58:53 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
`INFO[0000] Running RKE version: v1.0.0`
- Network plugin and version (if this is a network-related bug):
calico 3.10.1
- Others:

**Comment by huegelc** [comment]
@kubernetes/sig-network-bugs

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by huegelc** [comment]
nobody has this issue?

**Comment by Nurza** [comment]
I have the same issue with my clusters (wich are on different versions of k8s). I run weave for the CNI :

Are affected : 1.15.0, 1.15.2, 1.15.5, 1.17.1
Are not affected : 1.13.5, 1.17.3

Always in the default namespace.

**Comment by aojea** [comment]
/remove-triage unresolved

I'm not very familiar with rke but let's assume that it doesn't do any modification of the configuration parameters provided and kubernetes defaults.

Looking at your config, you have configured a cluster cidr with /24

 >   cluster_cidr: 10.2.225.0/24

and, by default, the kube-controller-manager uses a /24 per node

> --node-cidr-mask-size int32
> --
>   | Mask size for node cidr in cluster. Default is 24 for IPv4 and 64 for IPv6.

That means that the `pkg//controller/nodeipam/ipam/range_allocator.go` will only be able to assign one subnet from that range and the rest of the nodes.

For 5 nodes you should use minimum a mask 21 for the cluster_cidr if you want to keep the default node_cidr-mask-size, 2^(24-21) = 8

Although, I don't know if Calico has its own IPAM and the Kubernetes node ipam range allocator can be disabled, avoiding this messages.

**Comment by wyike** [comment]
> /remove-triage unresolved
> 
> I'm not very familiar with rke but let's assume that it doesn't do any modification of the configuration parameters provided and kubernetes defaults.
> 
> Looking at your config, you have configured a cluster cidr with /24
> 
> > cluster_cidr: 10.2.225.0/24
> 
> and, by default, the kube-controller-manager uses a /24 per node
> 
> > --node-cidr-mask-size int32
> >   | Mask size for node cidr in cluster. Default is 24 for IPv4 and 64 for IPv6.
> 
> That means that the `pkg//controller/nodeipam/ipam/range_allocator.go` will only be able to assign one subnet from that range and the rest of the nodes.
> 
> For 5 nodes you should use minimum a mask 21 for the cluster_cidr if you want to keep the default node_cidr-mask-size, 2^(24-21) = 8
> 
> Although, I don't know if Calico has its own IPAM and the Kubernetes node ipam range allocator can be disabled, avoiding this messages.

Hi @aojea, in our cluster using Antrea which uses kube-controller-manager to allocate podcidr; we already see error like :

> I1210 19:21:43.870293       7 controller_utils.go:185] Recording status change CIDRNotAvailable event message for node c141b605-7f1c-41c6-8bd1-1951827edea0
E1210 19:21:43.870341       7 controller_utils.go:260] Error while processing Node Add/Delete: failed to allocate cidr from cluster cidr at idx:0: CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range
I1210 19:21:43.870482       7 event.go:291] "Event occurred" object="c141b605-7f1c-41c6-8bd1-1951827edea0" kind="Node" apiVersion="v1" type="Normal" reason="CIDRNotAvailable" message="Node c141b605-7f1c-41c6-8bd1-1951827edea0 status is now: CIDRNotAvailable"

but kube-controller-manager keeps allocating pod ciders to all nodes and as a result, nodes have duplicate pod cidrs and cni adds duplicates route per the pod cidrs to different node.
Is this expected that kube controllelr -manager doesn't stop allocating while continue allocating duplicate pod cidrs to all nodes?

**Comment by aojea** [comment]
> but kube-controller-manager keeps allocating pod ciders to all nodes and as a result, nodes have duplicate pod cidrs and cni adds duplicates route per the pod cidrs to different node.
> Is this expected that kube controller-manager doesn't stop allocating while continue allocating duplicate pod cidrs to all nodes when CIDRNotAvailable happens?

please open a new issue and provide more detailed information, commenting in a closed issue will not provide visibility and there are also Antrea devs in the community that may help us to figure it out ...

> nodes have duplicate pod cidrs 
It would nice to explain your setup: controller manager configuration, what do you mean by duplicate cidrs? are the node.Spec.PodCIDR field with the same value in different nodes? ....

**Comment by aojea** [comment]
@wyike please tag me in the new issue ;)

**Comment by tnqn** [comment]
@aojea I reproduced the issue @wyike found with K8s v1.19.2. It's not specific to Antrea but an issue in NodeIPAMController that it could allocate duplicate PodCIDRs after the ClusterCIDR was exhaused if kube-controller-manager was restarted.
It was caused by incorrect count of `allocatedCIDRs` after restarting kube-controller-manager (same reason as https://github.com/kubernetes/kubernetes/issues/96416), so `allocatedCIDRs` could be greater than `maxCIDRs` and `AllocateNext` uses `s.allocatedCIDRs == s.maxCIDRs` to check if it's full then duplicate allocation could happen.
https://github.com/kubernetes/kubernetes/blob/e439c95e55522cac9fb743af761d68883c6f6269/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L154-L156

Since the incorrect count of `allocatedCIDRs` has been fixed by https://github.com/kubernetes/kubernetes/pull/96417, the duplicate allocation should be fixed by it as well. I confirmed K8s v1.19.5 and v1.20.0 don't have the issue.
